File: programmers/level2/괄호회전하기/괄호회전하기.java
import java.util.*;

class Solution {
    private String rotate(String s, int num) {
        String front = s.substring(num);
        String back = s.substring(0, num);
        return front+back;
    }
    
    public int solution(String s) {
        int answer = 0;
        
        for(int i = 0; i < s.length(); i++) {
            String rts = rotate(s, i);
            
            Stack<Character> st = new Stack<>();
            boolean flag = false;
            
            for(char c : rts.toCharArray()) {
                if (c == '(' || c == '{' || c == '[') st.push(c);
                else {
                    if (st.isEmpty()) continue;
                    
                    char peek = st.peek();
                    st.pop();
                    
                    if (peek == '(') flag = (c == ')') ? true : false;
                    else if (peek == '[') flag = (c == ']') ? true : false;
                    else if (peek == '{') flag = (c == '}') ? true : false;
                    
                    if (!flag) break;
                }
            }
            
            if (st.isEmpty() && flag) answer++;
        }
        return answer;
    }
}

File: programmers/level2/괄호회전하기/괄호회전하기.cpp
#include <string>
#include <vector>
#include <stack>
#include <iostream>
using namespace std;

int solution(string s) {
    int answer = 0;
    
    for (int i = 0; i < s.size(); i++) {
        string front = s.substr(i);
        string back = s.substr(0, i);
        string rts = front + back;
        
        bool flag = false;
        stack<char> st;

        for (auto& c : rts) {
            if (c == '(' || c == '[' || c == '{') st.push(c);
            else {
                if (st.empty()) continue;
                
                char top = st.top();
                st.pop();
                
                if (top == '(') {
                    flag = c == ')' ? true : false;
                } else if (top == '[') {
                    flag = c == ']' ? true : false;
                } else {
                    flag = c == '}' ? true : false;
                }
                
                if (!flag) break;
            }  
        }
        
        if (flag && st.empty()) answer++;
    
    }
    
    
    return answer;
}